File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-4.java
class Solution {
    HashMap<Integer, Integer> map = new HashMap<>();

    public int minCostClimbingStairs(int[] cost) {
        return Math.min(climbStairs(0, cost), climbStairs(1, cost));
    }

    public int climbStairs(int i, int[] cost) {
        if (map.containsKey(i)) {
            return map.get(i);
        }

        if (i >= cost.length) {
            map.put(i, 0);
            return 0;
        }

        if (i == cost.length - 1) {
            map.put(i, cost[i]);
            return cost[i];
        }

        int minCost = cost[i] + Math.min(climbStairs(i + 1, cost), climbStairs(i + 2, cost));
        map.put(i, minCost);
        return map.get(i);
    }
}

File: Data Structures & Algorithms/climbing-stairs/submission-4.py
class Solution:
    def climbStairs(self, n: int) -> int:
        d = {}
        def dfs(step: int) -> int:
            ways = d.get(step)
            if ways:
                return ways

            if step > n:
                return 0
            if step == n:
                return 1
            
            d1 = dfs(step + 1)
            d2 = dfs(step + 2)
            ways = d1 + d2
            d[step] = ways
            return ways
        
        return dfs(0)